File: week3/CountIncreasingPathsInGrid/countIncreasingPathsInGrid.cpp
#include <iostream> 
using namespace std;

int countPathsinGrid(vector<vector<int>> prid);
int walkOverGrid(int i, int j, int prev, vector<vector<int>> dp, vector<vector<int>> prid);

int main(){
    vector<vector<int>> input = {{1,1},{3,4}};
    int output = countPathsinGrid(input);
    cout << output;
}

int countPathsinGrid(vector<vector<int>> prid){
    int m = prid.size(), n = prid[0].size(), totalPath = 0;
    vector<vector<int>> dp (n, vector<int>(m,-1));
    
    for(int i=0; i<m;i++){
        for(int j=0;j<n;j++){
            totalPath+=walkOverGrid(i,j,-1,dp,prid);
        }
    }

    return totalPath;
}

int walkOverGrid(int i, int j, int prev, vector<vector<int>> dp, vector<vector<int>> prid){
    int total=1;
    // checking no over range
    if(i<0 || i >= prid.size() || j<0 || j >= prid[0].size() || prid[i][j] <= prev){
        return 0;
    }
    // chekc the cell already visted
    if (dp[i][j] != -1){
        return dp[i][j];
    }
    // walk over diff direction
    if (dp[i][j] == -1){
        vector<pair<int,int>> direction = {{1,0},{-1,0},{0,1},{0,-1}};
        for(auto &dir : direction){
            total+=walkOverGrid(i+dir.first, j+dir.second, prid[i][j], dp, prid);
        }
    }
    dp[i][j] = total;
    // return with mod
    int mod = 1e9+7;
    return int(total%mod);
}

File: week3/CountIncreasingPathsInGrid/countIncreasingPathsInGrid.py
class Solution:
    @classmethod
    def countPathsinGrid(self, grid: list[list[int]])->int:
        m = len(grid)
        n = len(grid[0])
        paths = 0       
        dp = [[-1] * n for x in range(m)]
        
        def walk_over_grid(i: int, j: int, prev: int, checking: int):
            cout = 1
            # make sure it's not out of range
            if i < 0 or i >= m or j < 0 or j >= n or prev >= grid[i][j]:
                return 0
            if dp[i][j] != -1:
                return dp[i][j]
            # only checking unvisted cell
            directions = [[1,0], [-1,0], [0,1],[0,-1]]
            if dp[i][j] == -1:
                for dir in directions:
                    cout += walk_over_grid(i+dir[0], j+dir[1], grid[i][j], checking+1)
            dp[i][j] = cout
            return cout

        for i in range(m):
            for j in range(n):
                paths += walk_over_grid(i, j, -1, 1)
    
        # return it modulo 10^9 + 7.
        mod = (int)(1e9+7)
        return paths % mod

# Input: grid = [[1,1],[3,4]]
# Output: 8

input = [[1,1],[3,4]]
output = Solution.countPathsinGrid(input)
print(output)

# Input: grid = [[1],[2]]
# Output: 3
input2 = grid = [[1],[2]]
output2 = Solution.countPathsinGrid(input2)
print(output2)